pkg/dyninst/decode/types.go

```diff
@@ -0,0 +1,960 @@
+	keyTypeID        ir.TypeID
+	keyTypeName      string
+	keyTypeSize      uint32
+	valueTypeID      ir.TypeID
+	valueTypeName    string
+	valueTypeSize    uint32
+	keyFieldOffset   uint32
+	valueFieldOffset uint32
+
+	// Internal Go swiss map representation fields
+	dirPtrOffset     uint32
+	dirPtrSize       uint32
+	dirLenOffset     uint32
+	dirLenSize       uint32
+	usedOffset       uint32
+	usedSize         uint32
+	ctrlOffset       uint32
+	ctrlSize         uint32
+	slotsOffset      uint32
+	slotsSize        uint32
+	groupFieldOffset uint32
+	groupFieldSize   uint32
+	dataFieldOffset  uint32
+	dataFieldSize    uint32
+	tableTypeID      ir.TypeID
+	groupTypeID      ir.TypeID
+	groupSliceTypeID ir.TypeID
+	elementTypeSize  uint32
+}
+type goSwissMapGroupsType ir.GoSwissMapGroupsType
+type goChannelType ir.GoChannelType
+type goEmptyInterfaceType ir.GoEmptyInterfaceType
+type goInterfaceType ir.GoInterfaceType
+type goSubroutineType ir.GoSubroutineType
+type eventRootType ir.EventRootType
+
+// Compile-time type assertions
+var (
+	_ decoderType = (*baseType)(nil)
+	_ decoderType = (*pointerType)(nil)
+	_ decoderType = (*structureType)(nil)
+	_ decoderType = (*arrayType)(nil)
+	_ decoderType = (*voidPointerType)(nil)
+	_ decoderType = (*goSliceHeaderType)(nil)
+	_ decoderType = (*goSliceDataType)(nil)
+	_ decoderType = (*goStringHeaderType)(nil)
+	_ decoderType = (*goStringDataType)(nil)
+	_ decoderType = (*goMapType)(nil)
+	_ decoderType = (*goHMapHeaderType)(nil)
+	_ decoderType = (*goHMapBucketType)(nil)
+	_ decoderType = (*goSwissMapGroupsType)(nil)
+	_ decoderType = (*goChannelType)(nil)
+	_ decoderType = (*goEmptyInterfaceType)(nil)
+	_ decoderType = (*goInterfaceType)(nil)
+	_ decoderType = (*goSubroutineType)(nil)
+	_ decoderType = (*eventRootType)(nil)
+)
+
+func newDecoderType(
+	irType ir.Type,
+	types map[ir.TypeID]ir.Type,
+) (decoderType, error) {
+	switch s := irType.(type) {
+	case *ir.GoSwissMapHeaderType:
+		dirPtrField, err := getFieldByName(s.RawFields, "dirPtr")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		dirPtrOffset := dirPtrField.Offset
+		dirPtrSize := dirPtrField.Type.GetByteSize()
+		dirLenField, err := getFieldByName(s.RawFields, "dirLen")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		dirLenOffset := dirLenField.Offset
+		dirLenSize := dirLenField.Type.GetByteSize()
+
+		usedField, err := getFieldByName(s.RawFields, "used")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		usedOffset := usedField.Offset
+		usedSize := usedField.Type.GetByteSize()
+
+		slotsField, err := getFieldByName(s.GroupType.RawFields, "slots")
+		if err != nil {
+			return nil, fmt.Errorf("malformed swiss map header type: %w", err)
+		}
+		slotsFieldType, ok := types[slotsField.Type.GetID()]
+		if !ok {
+			return nil, errors.New("type map slot field not found in types: " + s.GroupType.Name)
```

Comment:
why not just `fmt.Errorf` here and elsewhere? I think of errors.New as really for static strings and don't see benefit for things like this but it's not a blocking comment